engine/src/main/java/org/pentaho/di/shared/SharedObjectsManager.java

```diff
@@ -0,0 +1,223 @@
+  private volatile boolean partitionSchema_initialized = false;
+  private volatile boolean clusterSchema_initialized = false;
+
+  /**
+   * Create an instance of SharedObjectManager using the Bowl's SharedObjectIO
+   * @param bowl
+   * @return
+   */
+  public static SharedObjectsManager getInstance( Bowl bowl ) {
+    return new SharedObjectsManager( bowl.getSharedObjectsIO() );
+  }
+  SharedObjectsManager( SharedObjectsIO sharedObjectsIO ) {
+    this.sharedObjectsIO = sharedObjectsIO;
+  }
+
+  /**
+   * Call the SharedObject type specific method to populate the internal map
+   * @param type The SharedObject type
+   * @throws KettleException
+   */
+  private void populateSharedObjectMap( String type ) throws KettleException {
+    switch ( SharedObjectType.valueOf( type.toUpperCase() ) ) {
+      case CONNECTION:
+        populateDbMetas();
+        break;
+      case SLAVESERVER:
+        populateSlaveServers();
+        break;
+      case PARTITIONSCHEMA:
+        // populate partitionSchemas
+        break;
+      case CLUSTERSCHEMA:
+        //populate clusterSchemas
+        break;
+      default:
+        log.error( " Invalid Shared Object type " + type );
+        throw new KettleXMLException( "Invalid shared object type " + type );
+    }
+  }
+
+  /**
+   * Populate the SharedObjectsMap. The map is of the form {type, Map{name, SharedObjectInterface}}.
+   * @throws KettleException
+   */
+  private void populateDbMetas() throws KettleException {
+    if ( !dbconn_initialized ) {
+      synchronized ( this ) {
+        if ( !dbconn_initialized ) {
+          Map<String, Node> nodeMap = sharedObjectsIO.getSharedObjects( DB_TYPE );
+          Map<String, SharedObjectInterface> dbSharedObjectMap = new HashMap<>();
+          for ( String name : nodeMap.keySet() ) {
+            SharedObjectInterface dbSharedObject = new DatabaseMeta( nodeMap.get( name ) );
+            if ( !dbSharedObjectMap.containsKey( name ) ) {
+              dbSharedObjectMap.put( name, dbSharedObject );
+            }
+          }
+          sharedObjectsMap.put( DB_TYPE, dbSharedObjectMap );
+        }
+      }
+      dbconn_initialized = true;
+    }
+  }
+
+  private void populateSlaveServers() throws KettleException {
+    if ( !slaveServer_initialized ) {
+      synchronized ( this ) {
+        if ( !slaveServer_initialized ) {
+          Map<String, Node> nodeMap = sharedObjectsIO.getSharedObjects( SLAVESERVER_TYPE );
+          Map<String, SharedObjectInterface> slaveServerMap = new HashMap<>();
+          for ( String name : nodeMap.keySet() ) {
+            SharedObjectInterface slaveServerSharedObject = new SlaveServer( nodeMap.get( name ) );
+            if ( !slaveServerMap.containsKey( name ) ) {
+              slaveServerMap.put( name, slaveServerSharedObject );
+            }
+          }
+          sharedObjectsMap.put( SLAVESERVER_TYPE, slaveServerMap );
+        }
+      }
+      slaveServer_initialized = true;
+    }
+  }
```

Comment:
This is a lot of copy-paste code. The only variables here are the type string and the constructor call. Those could be refactored out as arguments.